New Module: HTStream

bnjenner opened the pull request:
This PR is for a module supporting HTStream (https://github.com/s4hts/HTStream), a suite of preprocessing tools for high throughput sequencing data. HTStream is the main preprocessing tool used by the UC Davis Bioinformatics Core for their analyses and workshops, and this module has played an important role in improving sample QC workflows in both settings.

Many thanks for the thorough documentation; my supervisors and I hope to be in contact soon!

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` has been updated
 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository or attached to this PR
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
 - [x] Module does not do any significant computational work

ewels commented on .github/CONTRIBUTING.md:
```diff
@@ -6,7 +6,11 @@ I try to manage the required tasks for MultiQC using GitHub issues, you probably
 
 However, don't be put off by this template - other more general issues and suggestions are welcome! Contributions to the code are even more welcome ;)
 
+<<<<<<< HEAD
+> _If you need help using MultiQC then the best place to go is the Gitter chatroom where you can ask me questions directly: https://gitter.im/ewels/MultiQC_
+=======
 > _If you need help using MultiQC then the best place to go is the Gitter chatroom where you can ask me questions directly: [https://gitter.im/ewels/MultiQC](https://gitter.im/ewels/MultiQC)_
+>>>>>>> upstream/master
```
Whenever you see `<<<<<<< HEAD`, `=======` and `>>>>>>> upstream/master`, this is a git merge conflict marker that has been incorrectly committed instead of being resolved.

These need to be removed - see https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line

Have a look at the diff in the _"Files changed"_ tab to see what code is being changed and make sure that it's only the stuff that you intend to be adding.

Let me know if you need any help!

ewels commented on multiqc/modules/htstream/apps/AdapterTrimmer.py:
```diff
@@ -0,0 +1,167 @@
+			html += '<div class="alert alert-info"> <strong>Notice:</strong> No adapters were trimmed from samples. </div>'	
+			return html
+
+		# If too many samples, forget about it.
+		if len(json.keys()) > 150:
```
Any specific reason for this?

Generally your module shouldn't make decisions like this - firstly, MultiQC has a bunch of mechanisms in place to make large bargraphs usable. Second, a lot of people use the data export files generated from MultiQC for downstream analysis even if the plots in the report are unreadable. These files are not generated if you exit early like this.

ewels reviewed (changes requested):
Hi @bnjenner,

Many thanks for this pull-request! I haven't had a full test of it yet but have spotted a few things from a speed skim - see comments.

In addition to the specific comments below, I would like to make the slightly exceptional request for you to use [Python Black](https://black.readthedocs.io/en/stable/installation_and_usage.html) on your new code:

```bash
black -l 120 -t py38 multiqc/modules/htstream/
```

(NB: _just_ your code and not anything else in the MultiQC package, yet)

Black handles Python code formatting and this command will somewhat drastically reformat your files. This is good for me as a maintainer, as Black forces all code to use the same code formatting style.

I'm not yet using Black in MultiQC and this is not mentioned anywhere yet, but I will soon. In your case it will overwrite basically all of the code you have written and I want to make sure that the git history shows the code as yours and not mine :) Additionally, it will help me to review the code as it I find it easier to read.

I hope this is ok, let me know how it goes. If in doubt I can run it for you.

Phil

ewels commented:
I updated your PR and fixed all of the merge conflict mess. I wasn't able to push this directly into this PR so I have made a PR to your fork to apply the fixes: https://github.com/s4hts/MultiQC/pull/4

Please merge that and then we can continue this PR review 🚀 

Phil

ewels commented on CSP.txt:
```diff
@@ -46,5 +46,15 @@ script-src 'self'
     'sha256-8kPZc83fQr/qzo5ISaYAcLl4M5meLuuTzHFVRGb4GhU=' # multiqc_mpl.js
     'sha256-1pWjCVoi5AoQvT52zb72qGCpIpfJzlc9trKcbcarXUY=' # multiqc_toolbox.js
     'sha256-gB0osLOAZOyj1tcLhUs/xzf98qK3dfViuIJdX0lI8xA=' # multiqc_fastqc.js
+    
+    'sha256-TUYHIjQsABv4n1G4GdIAz0ZljvuXPEvQ9T20Sok7TvE=' # ////////////////////////////////////////////////// Base JS for MultiQC Reports//
+    'sha256-teMLGfDW72TRam2f0Fnj53uchjf3tQR9N0xo+hR2PXU=' # ////////////////////////////////////////////////// MultiQC Table code///////////
+    'sha256-AEcY37NW7iIjmlzDOPZvuAorKHxXZxsECPnuSP52fB8=' # ////////////////////////////////////////////////// HighCharts Plotting Code/////
+    'sha256-iPIReyQyAoheerXqjtulr8WN7lPMWacMVU4awy5CTJg=' # ////////////////////////////////////////////////// Static MatPlotLib Plots Javas
+    'sha256-KGgqQTL/PWbF29mNXwFnepbDYBEnTKdMtejlNhEiXfs=' # ////////////////////////////////////////////////// MultiQC Report Toolbox Code//
+    'sha256-RO0nmf6TEJtrp+R4JF8N4UrF+A2Hfkb8BKxyS7nM5go=' # // Return JS code required for plotting a single sample// RSeQC plot. Attempt to
+    'sha256-kH3IklvmfvBiZprdCgEZqGuZjhiGEF5V3tHncoRciQI=' # // Javascript for the FastQC MultiQC Mod///////////////// Per Base Sequence Cont
+    'sha256-NXm6btt3SaSrMpn3ubq276L3yuxUtGJ8i6ds0Wa8GTA=' # // Functions//////////////////////////////////////////////////// Button Status S
+
```
I don't think that you should need these any more since the failing CI was fixed upstream (which you've now pulled in). Could you please try removing this and see if the CI passes?

It might be that you need two new lines for the new module-specific JS + CSS imports.

ewels reviewed (commented):
Many thanks for this module @bnjenner - it's a huge volume of work and very impressive. Thank you also for being so patient with my slow review process..

As it's so much code it's difficult to review very thoroughly, so I'm going to do this in a few steps:

* [x] Check that the CI tests pass (thanks for the super speedy updates)
* [ ] Review just the report output, not the code
* [ ] Brief read over the code (not expecting any big surprises given what I've skimmed so far)

bnjenner commented:
My mistake, hadn't realized I did not implement the fix for all the files! And no problem, I appreciate the patience on your end as well. Thank you for your cooperation and for taking the time to review the code!

ewels commented:
Ok, first round of thoughts after looking at a report from the test data.. I'll do a numbered list so that it's easy to respond to each point.

1. Can we get rid of the `hts_` prefixes? I'm not sure that they add much.
2. _Preprocessing Statistics_: It seems that the y-value has a different unit for most categories. Honestly, I'm not sure how useful this plot is. I wonder if it would be better to just use the _General Statistics_ table for these key values?
3. _Fragment Reduction_ and _Basepair Reduction_: my gut feeling is that these would be better as a single chart with multiple datasets, as it's basically different takes on the same thing.
4. _hts_Stats_: The first two tables again feel like they belong in _General Statistics_ really. Can hide certain columns by default - eg. it's typical for `%N` to be hidden by default in other modules, and `%SE` can be hidden if all the same value (which will be the case most of the time).
5. _Base by Cycle: Single End_: The _Base by cycle_ dataset is really nice, but I generally try to avoid having datasets as samples if at all possible. Here you have not many samples and they have short sample names so the buttons look fine. It will be very common for users to have many many more samples with much longer names and the buttons just don't scale. Could you set up the plot to show a single sample when the line on the main plot is clicked? I did this in the _RSeQC Junction Saturation_ plot for example. It's not ideal and requires a bit of additional custom code but maintains the data representation in a scalable manner. Alternatively you could steal the FastQC _Per Base Sequence Content_ heatmap as other modules have done (x-ref https://github.com/ewels/MultiQC/issues/1102)
6. _Quality by Cycle: Single End - Heatmap_: Same issue as above with dataset buttons. Also the y-axis doesn't have a label. Whilst having this as a dataset is very cool (I must admit I did a 😮 when I clicked) I wonder if this is beyond scope for MultiQC. I chose to exclude certain content similar to this from FastQC output. MultiQC was always intended as a summary only with end users encouraged to dig through to the underlying data and single-sample report if something looks unexpected.
7. _SeqScreener_: I guess that these numbered sections are generated for each contaminant type? Instead of having separate tables and sections like this, I wonder if we could have a stacked bar graph with a category for each type? The numbers could theoretically sum to > 100% presumably, but that shouldn't be an issue if we warn about that in the plot description. I think this would scale better, be more condensed and also make it easier to spot significant contaminants between the different categories. Could have two datasets - reads and calculated percentages (numeric, so could sum to > 100%). Happy to also consider alternatives if you have a better idea here.
8. _hts_AdapterTrimmer_ and _hts_QWindowTrim_ tables: Not sure that the notes column adds much here? Also the `%BP lost` is already plotted in the _Basepair Reduction_ plot and the remaining columns might be better suited to _General Statistics_
9. _AdapterTrimmer: Trimmed Basepairs Composition_: Maybe disable the _Percentages_ view if we only have one category (maybe I should add that logic into core MultiQC code.. 🤔). To be honest, when a single category (most of the time..?) it's basically identical to the table column above so I wonder if we can drop this plot..
10. _QWindowTrim: Read Composition of Bps Trimmed_: As above. If keeping these plots I wonder if we can combine them and have them as two datasets, as they're very similar..?
11. _hts_PolyATTrim_, _hts_LengthFilter_: Ok as I scroll I realise that there are loads of different types of trimming and filtering steps 😅 Can these all be combined into a single report section / set of plots / table? It would be nice to be able to compare how much was removed in each step. I find it a bit overwhelming to scroll past so many numbers, which is usually a warning sign for me. We want to be able to scan a MultiQC report and spot outliers / patterns quickly.
12. _hts_Stats 2_: I get the logic of having this at the end, but I wonder if it's better to combine with the first Stats section with multiple datasets. Again so that it's easier to compare the before and after picture without having to scroll miles in the report.
13. _Read Lengths: Single End heatmap_: Does this need to be a heatmap? I feel like other modules have done this with line graphs before which feels a bit easier to interpret / compare to me.

Generally, as you'll probably get from the comments above, I wonder if we can compress the report output quite a lot. This will hopefully make the report easier for end-users to read and digest, and also potentially remove the need for sub-sub sections (eg. multiple plots per section) which I try to avoid if at all possible.

Lots to go through there. Let me know what you think and of course argue the point if you think that certain stuff is better off the way that it is.

Thanks,

Phil

ewels commented:
Hi @bnjenner!

Did you mean to close this pull request, or was it accidental?

Phil

bnjenner commented:
I believe it was accidental! Thanks for bringing this to my attention!

On Sun, Mar 12, 2023 at 1:42 PM Phil Ewels ***@***.***> wrote:

> Hi @bnjenner <https://github.com/bnjenner>!
>
> Did you mean to close this pull request, or was it accidental?
>
> Phil
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ewels/MultiQC/pull/1303#issuecomment-1465294905>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMME3Q6VFUWAJNKPJ7R3QOLW3YYL7ANCNFSM4R4ZG3VQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bnjenner commented:
Hello Dr. Ewels, I seemed to have failed the PR on filtering the filenames step, but the error I received suggested the issue was not with my code. I may be mistaken but either way, this error does not seem to reference any of my files and instead the program ' bclconvert'. Please let me know if there is another issue or if the PR is good to go. Thanks again! 

`|           multiqc | Search path : /home/runner/work/MultiQC/MultiQC/test_data/data/modules
|         searching | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 2552/2552  
|            isoseq | Found 0 cluster reports
|            isoseq | Found 0 refine reports
|            isoseq | Found 0 refine reports
|gene_body_coverage | Empty geneBodyCoverage file found: empty_failed.geneBodyCoverage.txt
|junction_saturation | Junction saturation data all zeroes, skipping: 'all_zeroes'
|           compare | Found a 'labels' file expected by Sourmash: 'test_data/data/modules/sourmash/compare/no_matrix.labels.txt', however, could not find a accompanying matrix binary file 'test_data/data/modules/sourmash/compare/no_matrix'. So assuming that wasn't a Sourmash result
|      fastq_screen | Couldn't find number of reads with no hits for 'LIB014528_GEN00035295_TGACCA_L001_R2_screen'
|      fastq_screen | Couldn't find number of reads with no hits for 'LIB014528_GEN00035295_TGACCA_L001_R1_screen'
|           afterqc | Could not parse AfterQC JSON: 'broken.fq.gz.json'
|             fastp | Could not parse sample name from the fastp command:
fastp --json HG001.fastp.json --html HG001.fastp.html --report_title HG001 fastp report --thread 16 --detect_adapter_for_pe
Falling back to extracting it from the file name: "bad_command.fastp.json" -> "bad_command"
|            fastqc | Couldn't read 'fake_report_fastqc.zip' - Bad zip file
|        bclconvert | Found RunInfo.xml file in test_data/data/modules/bclconvert/3.9.3/merged/Reports but no Demux Stats file, skipping
|        bclconvert | Found multiple sequencer runs
|          whatshap | Could not find the 'ALL' chromosome for GM24385_0_1x_full_opt_trunc. Make sure that the input is not truncated. MultiQC will summarize information for all available chromosomes instead.
|           multiqc | Report      : all_ignored.html
|           multiqc | Data        : all_ignored_data
|           multiqc | MultiQC complete
No report should have been generated! Likely due to a module missing self.ignore_samples() or ModuleNoSamplesFound when no matching files are found
Error: Process completed with exit code 1.`

bnjenner commented:
Hello Dr. Ewels, I have spent some time polishing this module and updating it for your new checks. The module has since passed all the formatting and unit tests provided by you on github. Are we good to merge this PR? :)

bnjenner reviewed (commented):
black is no longer used in favor for ruff

bnjenner commented:
So v1.22 seems to have bricked my implementation of this module. 

In order to include the functionality we wish to have with this, I relied on adding the output of the plotting functions (which from my understanding used to just be html) to an html string that I could use in add.section().  I assume you guys made this change for a good reason, so I am gonna go back to the drawing board here. I would like to contribute to this project and adhere to your philosophy the best I can, so any advice you have for how to handle this situation would be very helpful, possibly a meeting over zoom if you have the time. 

Best wishes.

bnjenner commented:
Hello again, 

I refactored this module with the goal of better adhering to this projects coding philosophy.  It does not do anymore javascript trickery and should be more resilient to updates in the future. As a result, some of the functionality I had in the previous version of this module was removed (specifically, having a base by cycle graph for each individual sample similar to FastQC). In the future, I would like to re-implement what I removed, as it is an important aspect of HTStream, but I would only do so after the MultiQC maintainers decide what is most appropriate. 

Let me know what you all think, I would be happy to answer questions and begin working revisions.

vladsavelyev commented:
Sorry for late responses, and thanks so much for your patience!

Indeed, with the switch of the plotting backend, most of custom javascript was broken. Until we figure out a more maintainable and robust way to customise interactive plots in modules, we will keep it simple.

I'll give it another look. Since I can't push to the master branch of your fork, I'll have to create a new branch to push any fixed from my side.

bnjenner commented:
Hello again,

Thank you for your response. I am available all week to answer any
questions or squash any bugs. I look forward to hearing your feedback.

Best from California,

Bradley Jenner

On Sun, Jul 21, 2024 at 8:34 AM Vlad Savelyev ***@***.***>
wrote:

> Sorry for late responses, and thanks so much for your patience!
>
> Indeed, with the switch of the plotting backend, most of custom javascript
> was broken. Until we figure out a more maintainable and robust way to
> customise interactive plots in modules, we will keep it simple.
>
> I'll give it another look. Since I can't push to the master branch of your
> fork, I'll have to create a new branch to push any fixed from my side.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MultiQC/MultiQC/pull/1303#issuecomment-2241681423>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMME3Q6DZN7DS3EY7KQGIWLZNPIIJAVCNFSM4R4ZG3V2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMRUGE3DQMJUGIZQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

bnjenner commented:
Hello again,

Hope all is well.

Just checking in to see how the process is going. I imagine you're quite
busy, so I am here to help if you need to delegate some work.

Best,

Bradley Jenner

On Mon, Jul 22, 2024 at 9:15 AM Bradley Jenner ***@***.***> wrote:

> Hello again,
>
> Thank you for your response. I am available all week to answer any
> questions or squash any bugs. I look forward to hearing your feedback.
>
> Best from California,
>
> Bradley Jenner
>
> On Sun, Jul 21, 2024 at 8:34 AM Vlad Savelyev ***@***.***>
> wrote:
>
>> Sorry for late responses, and thanks so much for your patience!
>>
>> Indeed, with the switch of the plotting backend, most of custom
>> javascript was broken. Until we figure out a more maintainable and robust
>> way to customise interactive plots in modules, we will keep it simple.
>>
>> I'll give it another look. Since I can't push to the master branch of
>> your fork, I'll have to create a new branch to push any fixed from my side.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MultiQC/MultiQC/pull/1303#issuecomment-2241681423>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AMME3Q6DZN7DS3EY7KQGIWLZNPIIJAVCNFSM4R4ZG3V2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMRUGE3DQMJUGIZQ>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>

bnjenner commented:
@vladsavelyev  Thanks for your patience with this process. Hope we can chat about this PR soon. Enjoy the rest of your weekend.

bnjenner commented:
Hello again. 

I see this project is undergoing some feature additions, which explains why the only test that now fails involves a username and password for docker. 

Understandably, this PR is probably a low priority for the multiqc team at the moment, but please know that I am able and willing to help expedite this process.

If there is anything at all I can do to contribute so this module can be added to the main fork, please let me know.

bnjenner commented:
Hello again @vladsavelyev  and @ewels, 

Just inquiring on the status of this PR and it's expected merge timeline. 

Hope all is well.
